pkg/server/proto/server.proto

```diff
@@ -5402,3 +5407,44 @@ message InlineKeepalive {
   // that message to be confused with an inline keepalive
   string signature = 100000000;
 }
+
+// https://github.com/hashicorp/cloud-api/blob/master/hashicorp/cloud/common/pagination.proto
+// PaginationRequest are the parameters for a paginated list request.
+message PaginationRequest {
+  // The max number of results per page that should be returned. If the number
+  // of available results is larger than `page_size`, a `next_page_token` is
+  // returned which can be used to get the next page of results in subsequent
+  // requests. A value of zero will cause `page_size` to be defaulted.
+  uint32 page_size = 1  [json_name="page_size"];
+
+  // Specifies a page token to use to retrieve the next page. Set this to the
+  // `next_page_token` returned by previous list requests to get the next page of
+  // results. If set, `previous_page_token` must not be set.
+  string next_page_token = 2 [json_name="next_page_token"];
+
+  // Specifies a page token to use to retrieve the previous page. Set this to
+  // the `previous_page_token` returned by previous list requests to get the
+  // previous page of results. If set, `next_page_token` must not be set.
+  string previous_page_token = 3 [json_name="previous_page_token"];
+}
+
+// PaginationResponse is the response holding the page tokens for a paginated
+// list response.
+message PaginationResponse {
+  // This token allows you to get the next page of results for list requests.
+  // If the number of results is larger than `page_size`, use the
+  // `next_page_token` as a value for the query parameter `next_page_token` in
+  // the next request. The value will become empty when there are no more pages.
+  string next_page_token = 1 [json_name="next_page_token"];
+
+  // This token allows you to get the previous page of results for list
+  // requests. If the number of results is larger than `page_size`, use the
+  // `previous_page_token` as a value for the query parameter
+  // `previous_page_token` in the next request. The value will become empty when
+  // there are no more pages.
+  string previous_page_token = 2 [json_name="previous_page_token"];
+
+  // The total number of entities stored by the server.
+  // This field is an addition not included in the base hashicorp/cloud/common pagination proto
+  uint64 total_count = 3 [json_name="total_count"];
```

Comment:
Is this value necessary? Computing it is, strangely enough, a real drag on databases.

Reply:
We show these counts prominently on the “list” pages:

![total-count](https://user-images.githubusercontent.com/34030/202670656-a3494687-dcf6-43e4-af8b-b0238bc24a63.png)

Is the DB problem tractable?

Reply:
I'm thinking about in generically, like when there are 10k deployments. Would it be ok if the backend decided "this is too complex to compute" and just left it 0, such that if 0 then the frontend knows it's not calculated?

Reply:
Yeah, I think that’s reasonable. Or if the field is nullable in the proto, then we could set it to null to indicate “this can’t be computed for some reason”.

Reply:
Sounds fine to me, though I think we should decide whether we want to keep `total_count` on the generic `PaginationResponse` message, or whether we want to move it up to the specific `ListProjectsResponse` message (and all other `ListXResponse` messages that will support total count).

<img src="https://user-images.githubusercontent.com/34030/203288864-6886cc9a-a3fa-446e-b1be-a76d1aa8e5e7.png" width="320">

My rationale for including it on `PaginationResponse` is to give us a little nudge to consider it for endpoints such as this. But if we’re not always planning to support it then maybe it will set unhelpful expectations with consumers of the API.

Alternatively, we could indicate “total count not supported” using some other mechanism, like `oneof`…

```proto
message PaginationResponse {
  // ...
  oneof total_count_result {
    uint64 total_count = 3 [json_name="total_count"];
    bool total_count_not_supported = 4 [json_name="total_count_not_supported"];
  }
}
```

But this feels a little clunky, even to me 😅

Reply:
My vote would be for adding `count` to ListProjects, and not the global PaginationResponse. I don't think it's something we'll want to add to most list endpoints.

If I turn out to be wrong about that, and there turn out to be compelling reasons to add it broadly, wouldn't be bad toadd it to the `PaginationResponse` proto, and either keep filling or deprecate the rpc-specific `count` fields.

Reply:
I’m happy to proceed with that modification. I’ll add a note to the RFC.

Reply:
Mentioned this in Slack but posting here as well:

Was pairing w/ @jgwhite and he brought up another possibility, which is to separate count into its own method at the state layer, which can then be called where needed at the service layer. This prevents the ListProjects state method from having to `return []Projects, Pagination, count, err`.

I think this makes more sense and have updated the code/PR to reflect this.